### ISSUE TITLE
sql: panic on computed column when source INET column has no data  

### ISSUE THREAD

**Issue by robert-s-lee** [cockroachdb/cockroach#25670]
**BUG REPORT**

grabbing first octet of IPv4 into ip_a column.  Forgot put the IP value during a test which caused the panic.  

```
create table tbl1 (
  id uuid unique,       
  ip INET,       
  first int,     
  last int,     
  count int,        
  ip_a int as (string_to_array(host(ip),'.')[1]::int) stored,
  primary key (ip_a,first,id)
);

insert into tbl1 (id,first,last,count) values (b'kafef00ddeadbeed', 1,1,1);
```

```
*
* ERROR: [n1,client=[::1]:59272,user=root] a SQL panic has occurred while executing "INSERT INTO tbl1(id, first, last, count) VALUES (b'kafef00ddeadbeed', 1, 1, 1)": expected *DIPAddr, found *tree.DUuid
*
*
* ERROR: [n1,client=[::1]:59272,user=root] a panic has occurred!
*
panic while executing 1 statements: INSERT INTO _(_, _, _, _) VALUES (_, _, _, _); caused by expected *DIPAddr, found *tree.DUuid

goroutine 475 [running]:
runtime/debug.Stack(0x612d2c0, 0xc425121ec0, 0x3)
	/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
github.com/cockroachdb/cockroach/pkg/util/log.ReportPanic(0x612d2c0, 0xc425121ec0, 0xc420222000, 0x5cd67c0, 0xc4261ff440, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:195 +0xa5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc42589e000, 0x612d2c0, 0xc425121ec0, 0x5cf6440, 0xc4261f0120)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:521 +0x2d5
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func2(0xc42589e000, 0x612d2c0, 0xc425121ec0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:489 +0x61
panic(0x5cf6440, 0xc4261f0120)
	/usr/local/go/src/runtime/panic.go:505 +0x229
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.MustBeDIPAddr(0x612efc0, 0xc4261ccf60, 0xc4261ccf60, 0x612efc0, 0xc4261ccf60)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:1281 +0xfe
github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.glob..func23(0xc426f4ab00, 0xc420337560, 0x1, 0x1, 0x0, 0x0, 0xc42616fb80, 0x6123480)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/builtins.go:310 +0x6d
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval(0xc426d8be00, 0xc426f4ab00, 0xc426d8be00, 0x613a440, 0xc426d8be00, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3312 +0x31f
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval(0xc426d8bec0, 0xc426f4ab00, 0xc426d8bec0, 0x613a440, 0xc426d8bec0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3302 +0x13a
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*IndirectionExpr).Eval(0xc4261dacc0, 0xc426f4ab00, 0xc4261dacc0, 0x613a560, 0xc4261dacc0, 0xc426d62c58)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3201 +0x34c
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*CastExpr).Eval(0xc4261dad00, 0xc426f4ab00, 0x0, 0x0, 0x1, 0xc420337550)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:2701 +0x65
github.com/cockroachdb/cockroach/pkg/sql.GenerateInsertRow(0x0, 0x0, 0x0, 0xc4261e7180, 0x1, 0x5, 0xc4261ff350, 0xc4261d7b00, 0x5, 0x8, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/insert.go:443 +0x58c
github.com/cockroachdb/cockroach/pkg/sql.(*insertNode).Next(0xc42616fb80, 0x612d380, 0xc4261b49f0, 0xc42589e450, 0xc42589e3e0, 0x5c2bf20, 0xc42616fb01, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/insert.go:332 +0x2f8
github.com/cockroachdb/cockroach/pkg/sql.forEachRow(0x612d380, 0xc4261b49f0, 0xc42589e450, 0xc42589e3e0, 0x6123480, 0xc42616fb80, 0xc426d63318, 0xc42589e3e0, 0x6123480)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:2028 +0x5d
github.com/cockroachdb/cockroach/pkg/sql.countRowsAffected(0x612d380, 0xc4261b49f0, 0xc42589e450, 0xc42589e3e0, 0x6123480, 0xc42616fb80, 0x0, 0x1ee1c1b5, 0x10)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:2052 +0xd7
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithLocalEngine(0xc42589e000, 0x612d380, 0xc4261b49f0, 0xc42589e3e0, 0x2, 0x80d6b58, 0xc4261dc000, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:618 +0x31d
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc42589e000, 0x612d380, 0xc4261b49f0, 0x612f500, 0xc4261c1270, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:575 +0x8d5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc42589e000, 0x612d380, 0xc4261b49f0, 0x612f500, 0xc4261c1270, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:342 +0x9c7
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc42589e000, 0x612d380, 0xc4261b49f0, 0x612f500, 0xc4261c1270, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:83 +0x358
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc42589e000, 0x612d2c0, 0xc425121ec0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:911 +0x1d2b
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc420262e00, 0x612d2c0, 0xc425121ec0, 0x0, 0x0, 0xc425515028, 0x4, 0xc425515015, 0xd, 0x6115a00, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:491 +0xf9e
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func3(0xc420262e00, 0x612d2c0, 0xc425121ec0, 0xc42509f6c0, 0x5400, 0x15000, 0xc4207d9ed0, 0xc4202eb620, 0xc4203cd8e0, 0xc4202eb610)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:260 +0xfc
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:259 +0xf14

*
* ERROR: [n1,client=[::1]:59272,user=root] Reported as error 737f61c279984fd3b687c28607fa1d6b
*
```

$ cockroach version
Build Tag:    v2.0.1
Build Time:   2018/04/23 18:35:36
Distribution: CCL
Platform:     darwin amd64 (x86_64-apple-darwin13)
Go Version:   go1.10
C Compiler:   4.2.1 Compatible Clang 3.8.0 (tags/RELEASE_380/final)
Build SHA-1:  c92bc6e912880c28c0784d3208583fb27f482f76
Build Type:   release

**Comment by jordanlewis** [comment]
Ouch, this is going to be a bit irritating to fix. There are several interlocking problems. Another example of a crash is trying to insert a mis-typed value into the source of the computed column:

```
root@127.0.0.1:64831/demo> create table a (a inet, b string as(host(a)) stored);
CREATE TABLE

Time: 4.02572ms

root@127.0.0.1:64831/demo> insert into a values(1);
*
* ERROR: [n1,client=127.0.0.1:64833,user=root] a SQL panic has occurred while executing "INSERT INTO a VALUES (1)": expected *DIPAddr, found *tree.DInt
```

This is caused by the fact that we don't type-check inserts up front - we do it lazily. Mis-typed values are caught during encoding, all the way down in `MarshalColumnValue`.

Because of this, and because we start to evaluate the computed column expressions before actually marshalling values, we end up invoking builtins that expect their input types to be correct since normally the type checker would prevent mismatched types.

A similar problem exists for `NULL` - normally, the builtin infrastructure detects `NULL` arguments to builtins before evaluating them, simply returning `NULL` as a result.

The solution might be to type check `INSERT` up front, or to do a little more type checking of the computed column expressions before invoking them.

**Comment by jordanlewis** [comment]
Nevermind about the `NULL` problem - that's actually handled okay, as @knz also noticed in https://github.com/cockroachdb/cockroach/issues/25627#issuecomment-389984142.

I think this issue is actually a little scary as well, because if you look at the panic, you can see that the problem is that the types are wrong - that `host` is getting a `uuid` instead of an `inet`. But `host` is getting `ip`, which is in fact an `inet` type. I think there's probably an issue with the way the indexed vars are set up as well for computed columns - they're probably just looking by index, without regard to which columns are actually present in the insert.

**Comment by knz** [comment]
> if you look at the panic, you can see that the problem is that the types are wrong - that host is getting a uuid instead of an inet.

Wait what? where do you see a uuid?

**Comment by jordanlewis** [comment]
@knz, the first column of robert's sample table is a UUID. I have a fix out - see #25682.
